File: Study Plan/Top Interview 150/0125 Valid Palindrome/ValidPalindromeV2.java
/*  
    Problem
    --------------------------------------------------------------------------------
    125. Valid Palindrome
*/

/*  
    Solution
    --------------------------------------------------------------------------------
    - Runtime 3 ms (Beats 76.39% of users with Java)
    - Memory 42.85 MB (Beats 73.80% of users with Java)
*/

public class Solution {
    public boolean isPalindrome(String s) {
        if (s.length() == 1) return true;
        String phrase = s.toLowerCase();
        int left = 0;
        int right = phrase.length() - 1;
        while (left < right) {
            char first = phrase.charAt(left);
            char last = phrase.charAt(right);
            if (!Character.isLetterOrDigit(first)) {
                left++;
            } else if (!Character.isLetterOrDigit(last)) {
                right--;
            } else {
                if (first != last) {
                    return false;
                }
                left++;
                right--;   
            }
        }
        return true;
    }
}

File: Study Plan/Top Interview 150/0125 Valid Palindrome/ValidPalindromeV2.cs
/*  
    Problem
    --------------------------------------------------------------------------------
    125. Valid Palindrome
*/

/*  
    Solution
    --------------------------------------------------------------------------------
    - Runtime 57 ms (Beats 85.70% of users with C#)
    - Memory 43.74 MB (Beats 49.44% of users with C#)
*/

public class Solution
{
    public bool IsPalindrome(string s)
    {
        var phrase = s.ToLower();
        var left = 0;
        var right = phrase.Length - 1;

        while (left < right)
        {
            var first = phrase[left];
            var last = phrase[right];

            if (!char.IsLetterOrDigit(first))
            {
                left++;
            }
            else if (!char.IsLetterOrDigit(last))
            {
                right--;
            }
            else
            {
                if (first != last)
                {
                    return false;
                }

                left++;
                right--;
            }
        }

        return true;
    }
}